sdk/helper/keysutil/policy.go

```diff
@@ -1355,6 +1357,38 @@ func (p *Policy) Rotate(ctx context.Context, storage logical.Storage, randReader
 		}
 	}()
 
+	if err := p.RotateInMemory(randReader, false); err != nil {
+		return err
+	}
+
+	return p.Persist(ctx, storage)
+}
+
+// RotateInMemory rotates the policy but does not persist it to storage.
+// If restoreOnError is true and the rotation partially fails, the policy
+// state will be restored.
+func (p *Policy) RotateInMemory(randReader io.Reader, restoreOnError bool) (retErr error) {
+	if restoreOnError {
+		priorLatestVersion := p.LatestVersion
+		priorMinDecryptionVersion := p.MinDecryptionVersion
+		var priorKeys keyEntryMap
+
+		if p.Keys != nil {
+			priorKeys = keyEntryMap{}
+			for k, v := range p.Keys {
+				priorKeys[k] = v
+			}
+		}
+
+		defer func() {
+			if retErr != nil {
+				p.LatestVersion = priorLatestVersion
+				p.MinDecryptionVersion = priorMinDecryptionVersion
+				p.Keys = priorKeys
+			}
+		}()
+	}
+
 	if p.Keys == nil {
```

Comment:
I haven't tested this yet, but if you move this to L1485, you might be able to skip on the defer restore (though you'd still have to do it on `Rotate`).